torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1808,12 +1808,6 @@ void ProcessGroupNCCL::watchdogHandler() {
             }
             // signal the monitor thread on PG0 to start dumping
             shouldDump_.store(true);
-            // This sleep is used to give time for dumping before throwing
-            // exception
-            std::this_thread::sleep_for(
-                std::chrono::seconds(heartbeatTimeoutInSec_));
-            LOG(INFO) << logPrefix() << "slept for " << heartbeatTimeoutInSec_
-                      << " giving time for flight recorder dumps to finish.";
```

Comment:
hm

(a) i would tend to  agree with the motivation of exiting earlier, but worried about deleting a sleep that was ostensibly added to improve dump reliability.  Do we have confidence that removal does not impair dump reliability?
(b) i'm not sure what this sleep even does.  iiuc, this is the 'watchdog' thread, which does not perform a dump currently.  it just signals, and then the monitor thread performs the dump and has its own logic for waiting and exiting.  If that's true, this feels safe to remove, but i'm confused why it was here in the first place.  Was it added back at a time when we  did dump from this thread?

Reply:
I did induce a timeout with a test job and FR dumps happened as expected. The sleep was added in #123023 and it's hard to tell - just from reading the code on why it was added.
In that code, we already had a separate `heartbeatMonitorThread` and in this thread we do:
```
  // wait for the dump until timeout
  waitForFutureOrTimeout(
      asyncDebugDump,
      std::chrono::milliseconds(waitTimeoutDumpInMilSec_),
      "Flight recorder dump in heartbeatMonitor");
 ```
 From my reading of the code, and also by testing, the system appears to give FR enough time to complete the dumps without the sleep.
 The one concern is that maybe there are other monitoring services that take action when they see an exception thrown?

Reply:
cc @shuqiangzhang - do you remember why the sleep was necessary/or added in #123023 ?

Reply:
Without the sleep, The watchdog thread would throw an exception immediately after timeout/errors are detected

Reply:
Ah ok! Yeah, possibly because this was when everything was part of a single thread.
I tested the code after removing the sleep here:
Dump succeeded for my job where I induced a nccl timeout:
[https://fburl.com/mlhub/4jskge2q](https://l.facebook.com/l.php?u=https%3A%2F%2Ffburl.com%2Fmlhub%2F4jskge2q&h=AT17fJ9OIuWPqSttpKPoLRCMXgUP0tNoFRFDe-BETFH8-NhLEUuxPXALiYf6Rwpcn634k3PbNHIC827TDe8Vs0bchotacgFTI6MH-80aBO1E02oXi9-XR-uxUpATRdfb9UVbw8ooPXs)
Watchdog did successfully fire for this job too: https://fburl.com/mlhub/0mchynw7

Reply:
To be safe, I introduced an environment variable that can be used to quickly revert back to the old behavior. 
The intent is to remove this environment variable after some time (maybe one month or so).

Internally, this is protected by a knob that we can use to flip back to the old behavior dynamically.

NOTE: 
I'm defaulting to the new behavior (i.e. don't sleep after exception) - but I can default this to old behavior to be even safer?
Let me know @shuqiangzhang/ @wconstab